components/content/simple-editor/index.vue

```diff
@@ -0,0 +1,431 @@
+<script setup lang="ts">
+import { type Object3D } from 'three'
+import { TresCanvas } from '@tresjs/core'
+import { TransformControls } from '@tresjs/cientos'
+import { useElementSize, useRefHistory } from '@vueuse/core'
+
+import type { BoxSceneNode, CameraSceneNode, ConeSceneNode, SceneSettings, SphereSceneNode } from './types'
+
+interface State { 
+  sceneSettings: SceneSettings
+  sceneNodes: (CameraSceneNode | BoxSceneNode | SphereSceneNode | ConeSceneNode)[] 
+}
+const state = ref<State>({ sceneSettings: { width: 800, height: 600 }, sceneNodes: [
+  {
+    id: crypto.randomUUID(),
+    type: 'camera',
+    properties: {
+      fov: 40,
+    },
+    position: [ 0, 5, 50 ],
+    rotation: [ 0, 0, 0 ],
+    scale: [ 1, 1, 1 ],
+  },
+] })
+const history = useRefHistory(state, { deep: true })
+history.clear()
+
+const currentCameraName = ref('nav')
+const currentCamera = computed(() => ({ render: renderCamRef, nav: navCamRef }[currentCameraName.value]?.value))
+
+const selectedNodeId = ref<string | null>()
+watch(selectedNodeId, () => {
+}, { immediate: true })
+const selectedNode = computed(() => state.value.sceneNodes.find(i => i.id === selectedNodeId.value))
+const cameraNode = computed(() => state.value.sceneNodes.find(i => i.type === 'camera') as CameraSceneNode)
+
+type Mode = 'translate' | 'rotate' | 'scale'
+const currentMode = ref<Mode>('translate')
+const modes: Mode[] = ['translate', 'rotate', 'scale']
+
+type Space = 'world' | 'local'
+const currentSpace = ref<Space>('world')
+const spaces: Space[] = ['world', 'local']
+
+//#region TresJS
+const tresCanvasRef = shallowRef()
+const renderCamRef = shallowRef()
+const navCamRef = shallowRef()
+const cameraHelperRef = shallowRef()
+const canvasContainerRef = shallowRef()
+const sceneNodeRefs = ref<{ [sceneNodeId: string]: Object3D }>({})
+
+const setCamerasManualWatcher = watchEffect(() => {
+  if (renderCamRef.value != null && navCamRef.value != null) {
+    renderCamRef.value.manual = false
+    navCamRef.value.manual = false
+    setCamerasManualWatcher()
+  }
+})
+
+watchEffect(() => {
+  if (tresCanvasRef.value?.context != null && currentCamera.value != null) {
```

Comment:
(non blocking)
Why so many watchEffects.
All this could be inside one watchEffect, and add some comments to organization

Reply:
If it is not split, some unecessary calculation will be executed when it is not needed (all dependencies will be tracked in a single function). I also like to isolate the scope of dependencies and do as small of a task per watcheffect as possible to understand the purpose of the watcheffect(comments will probably be helpfull, will add).